xla/service/gpu/gpu_performance_model.h

```diff
@@ -39,6 +51,72 @@ class GpuPerformanceModel {
   // Writes estimated execution time to FusionBackendConfig.reification_cost.
   static void RecordEstimatedRunTime(HloInstruction* instruction,
                                      const GpuHloCostAnalysis* cost_analysis);
+  static absl::Duration ComputeTime(const GpuDeviceInfo& gpu_device_info,
+                                    int64_t n_flops, int64_t n_threads);
+
+  static absl::Duration ProducerInputAccessTime(
+      const GpuHloCostAnalysis* cost_analysis,
+      const GpuDeviceInfo& gpu_device_info, const HloInstruction* producer,
+      const HloInstruction* fused_consumer = nullptr);
+};
+
+class GpuPerformanceWithCollectiveModel : public GpuPerformanceModel {
+ public:
+  // Different algorithms that can be used to perform the collective.
+  enum CollectiveAlgo {
+    RING = 0,
+    TREE,
+  };
+
+  // Table for max system bandwidths.
+  // Index 1 is for well-known GPU link BWs, used for intra-node estimate.
+  static constexpr double low_latency_max_bandwidths[3] = {
+      39.0 /* Volta-N1 */, 87.7 /* Ampere-N1 */, 87.7 /* Hopper-N1 */
+  };
+
+  static constexpr double per_channel_max_ring_LL128_bandwidths[3] = {
+      20.0 /* Volta (N1/N2/N4) */,
+      20.0 /* Ampere (N1/N2/N4) */,
+      36.7 /* Hopper (N1/N2/N4) */,
+  };
+
+  // Nvlink bandwidth for different compute cap.
+  // Note this is per lane bandwidth.
+  static constexpr double sm60_nvlink_bw = 18.0;
+  static constexpr double sm70_nvlink_bw = 20.0;
+  static constexpr double sm80_nvlink_bw = 20.0;
+  static constexpr double sm90_nvlink_bw = 20.0;
+
+  // PCIE bandwidth
+  static constexpr double pci_bw = 12.0;
+
+  // Discount factor for ring algorithm
+  static constexpr double ring_algo_factor = 0.92;
+
+  // Different tiers for intra-node bandwidth.
+  static constexpr double intra_node_speeds[13] = {
+      40.0, 30.0, 20.0, 18.0, 15.0, 12.0, 10.0, 9.0, 7.0, 6.0, 5.0, 4.0, 3.0};
+  // SM90 has different bandwidths.
+  static constexpr double intra_node_speeds_sm90[9] = {
+      60.0, 40.0, 30.0, 24.0, 20.0, 15.0, 12.0, 6.0, 3.0};
+
+  // Maximum number of channels allowed by NCCL
+  static constexpr int64_t max_nchannels_ring = 16;
+
+  // ll128 is by default enabled for Volta, Ampere and Hopper, ll128 by default
+  // launches 640 threads.
+  static constexpr int64_t ll128_nthreads = 640;
```

Comment:
A general comment is that there are so many NCCL implementation details here and I wonder how things change between different NCCL versions. As an example, how much of this is correct on a different NCCL version than 2.18 and if those differences matter.

Reply:
Also, would be make sense for NCCL itself to provide some sort of estimation API for such use cases, so they can change it based on implementation changes. I know its outside the scope of this PR, but may be something to discuss in other forums

Reply:
We have been discussing with NCCL team regarding this recently. The summary is that they don't want to maintain APIs that do estimation, the reasoning is that if they change algorithms internally, these would all need to change. They suggest that we do something similar to some compute ops which is profiling collectives on every system that XLA will run with different tensor sizes and maintaining a database of the real costs. This means we have a huge combinations of profilings that we need to perform.
I'm trying to see if they can at least provide some accessor APIs for their internal parameters such as thread count or bandwidth limit for each algorithm. That will at least address parameter mismatch issue with different NCCL versions.